### PR TITLE
Reduce Coupling between RzCore and Tasks ##core

### PR DESCRIPTION
**Pull request by thestr4ng3r**
(More context in the comments)

This makes the low-level tasks code almost completely independent of RzCore. RzCore then puts a layer on top if the tasks to allow having a task-local cons context. This also introduces function tasks in addition to just command tasks.

### CODE REVIEW

**Comment by ret2libc** [comment]
Are you removing sandbox feature from everywhere?
Also, why do you want to decouple them? Do you intend to move tasks outside of RzCore?

**Comment by thestr4ng3r** [comment]
Yeah I think we should remove sandbox completely. I see no way of it being remotely safe.

Yes, the underlying task scheduling mechanism could be completely ripped out from core and moved to util. This will make it easier to get an overview over since task code is more self-contained, but more importantly, we can make other modules below core access the task api. This means we can for example move the breaking and yielding functionality completely from cons to tasks and have it local by passing a pointer to the scheduler to for example a complex analysis function to allow reentrant cooperative multitasking everywhere.

**Comment by thestr4ng3r** [comment]
But in the short run, my plan is to just be able to run arbitrary functions in tasks instead of being forced to us a command. This will make it possible to reintroduce async rz-ghidra in cutter.

**Comment by XVilka** [comment]
I think sandbox should be removed in a separate PR. Otherwise it's fine.

File: librz/core/cmd_tasks.c [hunk 1]
```diff
@@ -40,7 +41,10 @@ static int task_break(RzCore *core, int tid) {
 		eprintf ("This command is disabled in sandbox mode\n");
 		return -1;
 	}
-	if (tid) {
+	if (!tid) {
```
Comment (XVilka) [inline librz/core/cmd_tasks.c:44]: What about using `rz_return_val_if_fail()` instead?

File: librz/core/task.c [hunk 2]
```diff
@@ -625,12 +526,316 @@ RZ_API int rz_core_task_del (RzCoreTaskScheduler *scheduler, int id) {
...
+	}
+	const char *cmd;
+	if (task == core->tasks.main_task) {
+		cmd = "";
+	} else {
+		cmd = ((CmdTaskCtx *)task->runner_user)->cmd;
+	}
+	switch (mode) {
+	case 'j':
+		{
```
Comment (XVilka) [inline librz/core/task.c]: Can you please use PJ API while you are on it?

File: test/unit/test_task.c [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+	LocalCtx a_ctx = {
+		.shared = &shared_ctx,
+		.id = 1,
+		.want_to_consume = 3
+	};
+	RzCoreTask *a = rz_core_task_new (&sched, runner, local_ctx_free, &a_ctx);
+	int a_id = a->id;
+	rz_core_task_enqueue (&sched, a);
+	a = NULL; // ownership moved to the scheduler, don't touch anymore!
```
Comment (XVilka) [inline test/unit/test_task.c:85]: We will see this pattern often probably. Might be useful to create a macros or a function for this.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cmd_tasks.c (hunk 1 above)
Comment (thestr4ng3r) [reply librz/core/cmd_tasks.c:44]: It's not an assertion-level failure, the user could have passed 0, but that is the main task which should not be breaked.

File: librz/core/cmd_tasks.c [hunk 4]
```diff
@@ -77,6 +81,9 @@ RZ_IPI RzCmdStatus rz_tasks_delete_handler(RzCore *core, int argc, const char **
 		return RZ_CMD_STATUS_ERROR;
 	}
 	int tid = rz_num_math (core->num, argv[1]);
+	if (!rz_core_task_is_cmd (core, tid)) {
+		return -1;
```
Comment (ret2libc) [inline librz/core/cmd_tasks.c]:
```suggestion
		return RZ_CMD_STATUS_ERROR;
```

File: librz/core/cmd_tasks.c [hunk 5]
```diff
@@ -98,6 +105,9 @@ RZ_IPI RzCmdStatus rz_tasks_wait_handler(RzCore *core, int argc, const char **ar
 	if (argc == 2) {
 		tid = rz_num_math (core->num, argv[1]);
 	}
+	if (!rz_core_task_is_cmd (core, tid)) {
+		return -1;
```
Comment (ret2libc) [inline librz/core/cmd_tasks.c]:
```suggestion
		return RZ_CMD_STATUS_ERROR;
```

**Review by ret2libc — CHANGES_REQUESTED** [review]
Nice thing overall!! I follow xvilka saying that it would be better to use pj. Also, here or in a very next PR it would be great to have doc for the other API functions and maybe an overall description of tasks as used in rizin (the fact it is collaborative, etc.)

File: librz/core/task.c [hunk 6]
```diff
@@ -625,12 +526,316 @@ RZ_API int rz_core_task_del (RzCoreTaskScheduler *scheduler, int id) {
...
+		break;
+	}
+	}
+}
+
+RZ_API void rz_core_task_list(RzCore *core, int mode) {
+	RzListIter *iter;
+	RzCoreTask *task;
+	if (mode == 'j') {
+		rz_cons_printf ("[");
```
Comment (wargio) [inline librz/core/task.c]: Same here. PJ should be used.

**Review by wargio — CHANGES_REQUESTED** [review]

File: test/unit/test_task.c (hunk 3 above)
Comment (thestr4ng3r) [reply test/unit/test_task.c:85]: What pattern do you mean exactly?

**Review by ret2libc — APPROVED** [review]

**Pull request merged**